psi4/src/psi4/libfock/DFJCOSK.cc

```diff
@@ -0,0 +1,1012 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Psi4: an open-source quantum chemistry software package
+ *
+ * Copyright (c) 2007-2022 The Psi4 Developers.
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This file is part of Psi4.
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "jk.h"
+#include "psi4/libqt/qt.h"
+#include "psi4/libfock/cubature.h"
+#include "psi4/libfock/points.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/electrostatic.h"
+#include "psi4/libmints/mintshelper.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/integral.h"
+#include "psi4/liboptions/liboptions.h"
+#include "psi4/lib3index/dftensor.h"
+
+#include <vector>
+#include <map>
+#include <algorithm>
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+using namespace psi;
+
+namespace psi {
+
+SharedMatrix compute_numeric_overlap(std::shared_ptr<DFTGrid> grid, std::shared_ptr<BasisSet> primary) {
+
+    // DOI 10.1063/1.3646921, EQ. 9
+
+    int nbf = primary->nbf();
+    BasisFunctions bf_computer(primary, grid->max_points(), grid->max_functions());
+    auto S_num = std::make_shared<Matrix>("Numerical Overlap", nbf, nbf);
+
+    // This loop could be parallelized over blocks of grid points. However, the cost of the loop is
+    // so small (< 10 seconds for a 200 heavy atom system), parallelism isn't necessary
+
+    for (size_t bi = 0; bi < grid->blocks().size(); bi++) {
+
+        // grid points in this block
+        auto block = grid->blocks()[bi];
+        int npoints_block = block->npoints();
+        int nbf_block = block->local_nbf();
+        auto w = block->w();
+
+        // compute basis functions at these grid points
+        bf_computer.compute_functions(block);
+        auto point_values = bf_computer.basis_values()["PHI"];
```

Comment:
For objects as complex as a matrix, mark `const` when you can. I think this is also `&`-able.

Reply:
This is a `SharedMatrix`, not a `Matrix`. I was under the impression that `const auto &` saves you from expensive copies, which is only an issue when working with `Matrix` objects, not shared pointers. Is there still a reason to `const auto &`?

Reply:
Oops, my mistake.

This is one of the reasons I hate Psi's overuse of `shared_ptr`. In that case, the benefits to casting `const` and `&` are _there_ but both minor. The `const` means that we can't have this pointer point to a different object, but we can still change the object. The `&` means we create a new pointer. Using it is a touch cheaper, but it's negligible on the scale of the entire algorithm.

To mark the matrix itself as `const`, which was what I was hoping for, you'd need `std::shared_ptr<const Matrix>`, which would ruin the nice `auto`.